File: 프로그래머스/3/43164. 여행경로/여행경로.java
import java.util.*;

class Solution {
    public String[] solution(String[][] tickets) {
        Map<String, PriorityQueue<String>> destinations = new HashMap<>();
        List<String> answer = new LinkedList<>();
        
        for(String[] ticket : tickets) {
            destinations.computeIfAbsent(ticket[0], destination -> new PriorityQueue<>())
                .add(ticket[1]);
        }
        
        dfs(destinations, "ICN", answer);
        
        return answer.toArray(new String[0]);
    }
    
    private void dfs(Map<String, PriorityQueue<String>> destinations, String departure, List<String> answer) {
        while(destinations.containsKey(departure) && !destinations.get(departure).isEmpty()) {
            dfs(destinations, destinations.get(departure).poll(), answer);
        }
        
        answer.add(0, departure);
    }
}

File: 프로그래머스/3/43164. 여행경로/여행경로.kt
import java.util.*;

class Solution {
    fun solution(tickets: Array<Array<String>>): Array<String> {
        val destinations: MutableMap<String, PriorityQueue<String>> = mutableMapOf()
        val answer: MutableList<String> = mutableListOf()
        
        for (ticket in tickets) {
            destinations.computeIfAbsent(ticket[0]){ PriorityQueue() }
                .add(ticket[1])
        }
        
        fun dfs(departure: String) {
            while (destinations[departure]?.isNotEmpty() == true) {
                dfs(destinations[departure]!!.poll())
            }
            
            answer.add(0, departure)
        }
        
        dfs("ICN")
        
        return answer.toTypedArray()
    }
}